x/wasm/keeper/keeper.go

```diff
@@ -565,7 +567,30 @@ func (k Keeper) callMigrateEntrypoint(
 	prefixStoreKey := types.GetContractStorePrefix(contractAddress)
 	vmStore := types.NewStoreAdapter(prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(sdkCtx)), prefixStoreKey))
 	gasLeft := k.runtimeGasForContract(sdkCtx)
-	res, gasUsed, err := k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+
+	migrateInfo := wasmvmtypes.MigrateInfo{
+		Sender:          senderAddress.String(),
+		OldStateVersion: OldStateVersion,
+	}
+	var res *wasmvmtypes.ContractResult
+	var gasUsed uint64
+	var err error
+	if OldStateVersion == nil {
+		res, gasUsed, err = k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+		if err != nil {
+			if strings.Contains(err.Error(), "The called function args arity mismatch, expected 3 args") {
```

Comment:
I'm pretty sure this gas handling together with the `Contains` check for the error can be exploited.
Setup: contract has old `migrate` entrypoint and does a lot of work there, but panics with "The called function args arity mismatch, expected 3 args" at the end. The VM behavior would look like this:
1. `Migrate` is called and calls into the entrypoint. After that `gasUsed` is set to however much the contract used and `err` is something like "... panicked at 'The called function args arity mismatch, expected 3 args', src/contract.rs:169:5"
2. The error message passes both `err != nil` and the `Contains` check, so VM tries to call `Migrate2`, which fails at the arity check with very little `gasUsed`, completely discarding the previous `gasUsed`.

Solution: We should only add to `gasUsed` and subtract from `gasLeft` in between calls, or handle this by calling `k.consumeRuntimeGas(sdkCtx, gasUsed)` and `gasLeft = k.runtimeGasForContract(sdkCtx)`.

Reply:
I think that's solved by moving the arity check to the cosmwasm-vm. Now the type system of rust prevents such a trick - cosmwasm-vm won't call migrate second time if the contract panics with any msg string.